cpp/src/arrow/util/async_generator.h

```diff
@@ -60,6 +60,9 @@ namespace arrow {
 // Readahead operators, and some other operators, may introduce queueing.  Any operators
 // that introduce buffering should detail the amount of buffering they introduce in their
 // MakeXYZ function comments.
+//
+// A generator should always be fully consumed before it is destroyed.
+// A generator should not emit a terminal item until it has finished all ongoing futures.
```

Comment:
Here, "emit a terminal item" is more like "let a future terminate with an error or an end-of-stream marker"? or i guess "item (future) that will terminate the stream when resolved". The wording reads a little oddly to me because it feels like the generator can see the future

Reply:
Well I hope the generator can see the future because it returned it ;)

I can try and fix up the warning.  I agree there is probably some ambiguity whether `emit` means "returning the future" or "completing a future" (this statement assumes it means the latter).

Something like:

```
A generator should ensure that all ongoing futures have completed before it completes a future with a terminal (error or end-of-stream) result.  All futures completed after that result should be the end-of-stream result.
```

Reply:
Yeah, that makes sense to me. 

I meant 'seeing the future (of whether the future will complete)' which…I'll stop there before I get more confused.

Reply:
I ended up with:
```
// A generator should not mark a future complete with an error status or a terminal value
//   until all outstanding futures have completed.  Generators that spawn multiple
//   concurrent futures may need to hold onto an error while other concurrent futures wrap
//   up.
```